exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java

```diff
@@ -130,562 +145,248 @@ public IterOutcome innerNext() {
-            }
-            builder.setMinorType(outputMinorType);
+          List<TypeProtos.MinorType> types = Lists.newLinkedList();
+          types.add(leftField.getType().getMinorType());
+          types.add(rightField.getType().getMinorType());
+          TypeProtos.MinorType outputMinorType = TypeCastRules.getLeastRestrictiveType(types);
+          if (outputMinorType == null) {
+            throw new DrillRuntimeException("Type mismatch between " + leftField.getType().getMinorType().toString() +
+                " on the left side and " + rightField.getType().getMinorType().toString() +
+                " on the right side in column " + index + " of UNION ALL");
           }
-
-          // The output data mode should be as flexible as the more flexible one from the two input tables
-          List<DataMode> dataModes = Lists.newLinkedList();
-          dataModes.add(leftField.getType().getMode());
-          dataModes.add(rightField.getType().getMode());
-          builder.setMode(TypeCastRules.getLeastRestrictiveDataMode(dataModes));
-
-          outputFields.add(MaterializedField.create(leftField.getPath(), builder.build()));
+          builder.setMinorType(outputMinorType);
         }
-        ++index;
-      }
-
-      assert !leftIter.hasNext() && ! rightIter.hasNext() : "Mis-match of column count should have been detected when validating sqlNode at planning";
-    }
 
-    private void inferOutputFieldsFromSingleSide(final BatchSchema schemaForNames, final BatchSchema schemaForTypes) {
-      outputFields = Lists.newArrayList();
+        // The output data mode should be as flexible as the more flexible one from the two input tables
+        List<TypeProtos.DataMode> dataModes = Lists.newLinkedList();
+        dataModes.add(leftField.getType().getMode());
+        dataModes.add(rightField.getType().getMode());
+        builder.setMode(TypeCastRules.getLeastRestrictiveDataMode(dataModes));
 
-      final List<String> outputColumnNames = Lists.newArrayList();
-      final Iterator<MaterializedField> iterForNames = schemaForNames.iterator();
-      while(iterForNames.hasNext()) {
-        outputColumnNames.add(iterForNames.next().getPath());
-      }
-
-      final Iterator<MaterializedField> iterForTypes = schemaForTypes.iterator();
-      for(int i = 0; iterForTypes.hasNext(); ++i) {
-        MaterializedField field = iterForTypes.next();
-        outputFields.add(MaterializedField.create(outputColumnNames.get(i), field.getType()));
+        container.addOrGet(MaterializedField.create(leftField.getPath(), builder.build()), callBack);
       }
+      ++index;
     }
 
-    public List<MaterializedField> getOutputFields() {
-      if(outputFields == null) {
-        throw new NullPointerException("Output fields have not been inferred");
-      }
-
-      return outputFields;
-    }
+    assert !leftIter.hasNext() && ! rightIter.hasNext() : "Mis-match of column count should have been detected when validating sqlNode at planning";
```

Comment:
True? Even in the case of a `SELECT *`? Any other cases where the set of columns isn't known at plan time?

Reply:
This is also in the original code. (Seems git diff did not do a good job). I think it might need cleanup, but it's probably out of scope for this JIRA. 

1. https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java#L620

Reply:
Thanks for the explanation.